### ISSUE TITLE
Compiling ImmutableMapFactory leads to deep subtype check

### ISSUE THREAD

**Issue by DarkDimius** [lampepfl/dotty#943]

**Comment by odersky** [comment]
Not sure this is a real errors. Deep subtypes can arise, but they are generally rare. So to go ahead with this, we'd first have to analyze that the deep subtype check is unexpected. Unfortunately, ImmutableMapFactory.scala alone compiles fine without a deep subtype, so the matter will be hard to diagnose.

**Comment by odersky** [comment]
I could get a real stack overflow with this minimized test case:

    scala-scala/src/library/scala/collection/generic> dotc GenTraversableFactory.scala GenericTraversableTemplate.scala

**Comment by smarter** [comment]
I managed to reduce this to two lines that produce a stack overflow (although it might be a different bug that I accidentally stumbled upon):
```scala
class Foo[A]
class Bar[CC[X] <: Foo[CC[X]]]
```
Here's the stack trace: https://gist.github.com/anonymous/164f565f9444cb9c4371

**Comment by smarter** [comment]
In case you're wondering how I got there, start from:
```scala
abstract class GenTraversableFactory[CC[X] <: GenTraversable[X] with GenericTraversableTemplate[X, CC]]
trait GenericTraversableTemplate[+A, +CC[X] <: GenTraversable[X]] extends HasNewBuilder[A, CC[A] @uncheckedVariance]
trait HasNewBuilder[+A, +Repr] extends Any
```
Remove a bunch of stuff and replace things by what they extend to get:
```scala
abstract class GenTraversableFactory[CC[X] <: HasNewBuilder[X, CC[X]]
```
Then remove type parameters and variance annotations until you're left with my example in the previous post.

**Comment by smarter** [comment]
`CC[X]` is not recognized as F-bounded type in `class Bar[CC[X] <: Foo[CC[X]]]` because of:
```scala
            case _: RefinedType => false
              // Note: it's important not to visit parents of RefinedTypes,
              // since otherwise spurious #Apply projections might be inserted.
```
which was introduced here: https://github.com/lampepfl/dotty/commit/ef4ecc11674335e095745b06f49f3387a9693777
If we want to continue using our current hk scheme, we should stop repurposing `checkNonCyclic` to insert`#apply` and instead do it separately.

**Comment by smarter** [comment]
Here's my best attempt at minimizing this issue:
```scala
import scala.annotation.unchecked.uncheckedVariance

trait GT[A] extends GTT[A, GT]

trait HNB[A]
trait GTT[+A, DD[X] <: GT[X]] extends HNB[DD[A] @uncheckedVariance] // Can be any annotation and still crash

class GTF[CC[X] <: GT[X] with GTT[X, CC]]
{
  def foo[A]: CC[A] = ???
  def bar[A](f: A): CC[A] = ???

  def tabulate: CC[CC[Int]] = bar(foo)
}
```

**Comment by smarter** [comment]
After applying #976 the stack overflow is gone, but the following still lead to a deep subtype check:
```shell
./bin/dotc -Yno-deep-subtypes ./scala-scala/src/library/scala/collection/GenIterable.scala ./scala-scala/src/library/scala/collection/GenIterableLike.scala ./scala-scala/src/library/scala/collection/GenMapLike.scala ./scala-scala/src/library/scala/collection/GenTraversable.scala ./scala-scala/src/library/scala/collection/GenTraversableLike.scala ./scala-scala/src/library/scala/collection/IndexedSeqLike.scala ./scala-scala/src/library/scala/collection/Iterable.scala ./scala-scala/src/library/scala/collection/MapLike.scala ./scala-scala/src/library/scala/collection/generic/ImmutableMapFactory.scala
```
However, it compiles fine without `-Yno-deep-subtypes`, maybe the types are complicated enough here that having deep subtypes is OK?

**Comment by smarter** [comment]
Now completely fixed in #976 by tweaking `compareAlias`! :tada:
